presto-parquet/src/main/java/com/facebook/presto/parquet/predicate/TupleDomainParquetPredicate.java

```diff
@@ -264,92 +283,56 @@ public static Domain getDomain(Type type, DictionaryDescriptor dictionaryDescrip
-            return Domain.create(ValueSet.copyOf(type, values), true);
-        }
-
-        return Domain.all(type);
+        // TODO: when min == max (i.e., singleton ranges, the construction of Domains can be done more efficiently
+        return getDomain(columnDescriptor, type, values, values, true);
     }
 
-    private static void failWithCorruptionException(boolean failOnCorruptedParquetStatistics, String column, ParquetDataSourceId id, Statistics statistics)
-            throws ParquetCorruptionException
+    public static long asLong(Object value)
     {
-        if (failOnCorruptedParquetStatistics) {
-            throw new ParquetCorruptionException(format("Corrupted statistics for column \"%s\" in Parquet file \"%s\": [%s]", column, id, statistics));
+        if (value instanceof Byte || value instanceof Short || value instanceof Integer || value instanceof Long) {
+            return ((Number) value).longValue();
         }
-    }
 
-    private static <T extends Comparable<T>> Domain createDomain(Type type, boolean hasNullValue, ParquetRangeStatistics<T> rangeStatistics)
-    {
-        return createDomain(type, hasNullValue, rangeStatistics, value -> value);
+        throw new IllegalArgumentException("Can't convert value to long: " + value.getClass().getName());
     }
 
-    private static <F, T extends Comparable<T>> Domain createDomain(Type type,
-            boolean hasNullValue,
-            ParquetRangeStatistics<F> rangeStatistics,
-            Function<F, T> function)
+    private static class DictionaryValueConverter
     {
-        F min = rangeStatistics.getMin();
-        F max = rangeStatistics.getMax();
+        private final Dictionary dictionary;
 
-        if (min != null && max != null) {
-            return Domain.create(ValueSet.ofRanges(Range.range(type, function.apply(min), true, function.apply(max), true)), hasNullValue);
-        }
-        if (max != null) {
-            return Domain.create(ValueSet.ofRanges(Range.lessThanOrEqual(type, function.apply(max))), hasNullValue);
+        private DictionaryValueConverter(Dictionary dictionary)
+        {
+            this.dictionary = dictionary;
         }
-        if (min != null) {
-            return Domain.create(ValueSet.ofRanges(Range.greaterThanOrEqual(type, function.apply(min))), hasNullValue);
+
+        private Function<Integer, Object> getConverter(PrimitiveType primitiveType)
+        {
+            switch (primitiveType.getPrimitiveTypeName()) {
+                case INT32:
+                    return (i) -> dictionary.decodeToInt(i);
+                case INT64:
+                    return (i) -> dictionary.decodeToLong(i);
+                case FLOAT:
+                    return (i) -> dictionary.decodeToFloat(i);
+                case DOUBLE:
+                    return (i) -> dictionary.decodeToDouble(i);
+                case FIXED_LEN_BYTE_ARRAY:
+                case BINARY:
+                case INT96:
+                default:
```

Comment:
Throw error in case of `default`? It may cause unexpected behavior in future if the caller code has changed.

Reply:
yep, will fix